00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,238 @@
+
+#### Pickles proofs
+
+A pickles proof is formed of a Pallas proof and a partially-verified Vesta proof, which are related by the structure of their respective public inputs.
+
+When we say that the Vesta proof is 'partially-verified', we mean that the Pallas proof has already run stages 1 ('Absorbing the polynomial commitments') and 3 ('Verifying the opening proof') of the verification algorithm, except for computing the recursion polynomial commitment in stage 3. As part of the pickles recursion, we are able to 'check' the recursion polynomial commitment by its evaluations at a random point in the *next* round of recursion (i.e. in the next 'opening proof'), so that we only ever have to run 1 full IPA check.
+
+The Pallas proof requires all steps of verification to be run as part of the non-native verifier circuit, but any previous Pallas proofs (i.e. any previous pickles proofs) that we have recursed over will also be amortised into a single Pallas recursion polynomial commitment.
+
+The values shared between the stages are exposed from the public inputs of the pickles Pallas proof, as well as a hash of the 'pickles public input' to be used by the user circuits embedded in the pickles instance. The details of this are elided for brevity.
+
+### High-level description of the algorithm
+
+As inputs, we take a list of 'recursion challenges', which will have length `log2(domain_size)` for each of the 2 groups.
+
+For the Vesta proof, [`log2(domain_size) = 16`](https://github.com/MinaProtocol/mina/blob/8814cea6f2dfbef6fb8b65cbe9ff3694ee81151e/src/lib/crypto/kimchi_backend/pasta/basic/kimchi_pasta_basic.ml#L17), and for the Pallas proof, [`log2(domain_size) = 15`](https://github.com/MinaProtocol/mina/blob/8814cea6f2dfbef6fb8b65cbe9ff3694ee81151e/src/lib/crypto/kimchi_backend/pasta/basic/kimchi_pasta_basic.ml#L16).
+
+For each of these, we have to compute the polynomial `b_poly`, which is defined by
+```
+b_poly(x) = (1 + chal[-1] x)(1 + chal[-2] x^2)(1 + chal[-3] x^4)...
+```
+and can be easily computed using a standard circuit of size `domain_size` using the algorithm [here](https://github.com/o1-labs/proof-systems/blob/cfc829220b44c1122863eca0db411560b99d6c8e/poly-commitment/src/commitment.rs#L294).
+
+Once the coefficients of `b_poly` have been determined, we want to provably construct its polynomial commitment by computing the MSM formed by each coefficient and the commitment from the URS that reprepresents the corresponding `x^i`. Since the basis of the MSM is known and fixed, we can convert the 254-bit scalings into a series of `k`-bit scalings, by computing
+```
+x = x_0 + 2^k x_1 + 2^(2k) x_2 + ...
+```
+and computing the scaling `x G` in parts as
+```
+x_0 G + x_1 (2^k G) + x_2 (2^(2k) G) + ...
+```
+where each of the `2^i G` terms are known constants.
+
+From here, we can use a standard 'bucketing' trick with `2^k` buckets to avoid doing any doublings or scalings at all, requiring only `n*(254/k)+2^(2k)` curve point additions:
```

Comment:
it's 2 additions per bucket, expect for the zero bucket, so the precise number is $2^{k + 1} - 2$
```suggestion
From here, we can use a standard 'bucketing' trick with `2^k` buckets to avoid doing any doublings or scalings at all, requiring only `n*(254/k)+2^(k + 1)` curve point additions:
```

Reply:
The formula was altogether wrong... The correct number of additions is at most $3n$ since $n$ is the size of MSM and thus the maximum number of buckets.

Reply:
Your understanding of the protocol was wrong, not the formula :man_facepalming: Please check next time before changing things.